### ISSUE TITLE
AxiosError's toJSON skips response

### ISSUE THREAD

**Issue by KoltesDigital** [axios/axios#4836]
#### Is your feature request related to a problem? Please describe.

My projects use Axios, and they always do in a way that that promise rejection will end up being logged by [Pino](https://github.com/pinojs/pino), with the error attached. I noticed that the log message contains the request config, but not the response, which would be helpful to fix the error. I believe that error responses (i.e. whose status is not validated) do not contain actual payload that would flood the log, but should only contain details about the error, therefore it's fine to log it. And, for specific projects where that's not the case, users can still remove the response before the error is dumped, using some feature the logging framework should provide (for Pino one would use [custom serializers](https://github.com/pinojs/pino/blob/master/docs/api.md#serializers-object), perhaps [formatters](https://github.com/pinojs/pino/blob/master/docs/api.md#formatters-object) or [redact](https://github.com/pinojs/pino/blob/master/docs/api.md#redact-array--object) too).

#### Describe the solution you'd like

Add `this.response` into [`AxiosError.toJSON`](https://github.com/axios/axios/blob/665db731099314ef6128188a5657bae0a142640d/lib/core/AxiosError.js#L33).

I could have made a PR, but I feel there is reason why you haven't done it, so this issue is rather a question about that.

#### Describe alternatives you've considered

Logging takes place at the API entrypoint (e.g. in an Express middleware) and has no access to the Axios request. Therefore one alternative would be to catch every request [as your doc mentions](https://axios-http.com/docs/handling_errors) and throw a new error including the response. Obviously that would be factorized in a function so that the code overhead would be low, but it'd still be a burden to think about adding it whenever developers write new code.

#### Additional context

Thanks for your efforts!

**Comment by gustf** [comment]
Hi there, any news regarding this?

**Comment by KoltesDigital** [comment]
Yes. I'm reimplementing my services in Rust. It perfectly solves this issue and many other ones.

Maybe actually making the PR would catch the maintainers' attention.

**Comment by gustf** [comment]
Sounds like fun rewriting in Rust! 👍 

And yes, you are probably right about making a PR

**Comment by KoltesDigital** [comment]
Cynicism aside, after I created this ticket, I've experimented with interceptors, and I think they can do the job pretty well. The idea is to transform the error in order to provide a custom serialization. For instance, the following code adds the response's body to the error when serialized:

```js
class AxiosErrorWrapper {
  constructor(error) {
    this.error = error;
  }

  toJSON() {
    return {
      ...this.error.toJSON(),
      responseData: this.error.response && this.error.response.data,
    };
  }
}

axios.interceptors.response.use(undefined, (error) => {
  if (error instanceof AxiosError) {
    return Promise.reject(new AxiosErrorWrapper(error));
  }
  return Promise.reject(error);
});
```

Logging the full response may be too much: there are lots of internal states and constants (!) that surely aren't worth it.

**Comment by meteorlxy** [comment]
For pino custom serializer:

```ts
       serializers: {
          err: pino.stdSerializers.wrapErrorSerializer((error) => {
            if (error.name === 'AxiosError') {
              const toJSON = error.toJSON.bind(error);
              error.toJSON = () => ({
                ...toJSON(),
                response: {
                  data: error.response?.data,
                  headers: { ...error.response?.headers },
                },
              });
            }
            return error;
          }),
        }
```

**Comment by ishowta** [comment]
I checked with git blame and it seems to be to prevent circular references during stringify.

https://github.com/axios/axios/commit/6b44e80adebf66984b7ad41fd71fcaac28f3fed8

**Comment by jasonsaayman** [comment]
closing as i agree interceptors should be able to do this

**Comment by KoltesDigital** [comment]
@jasonsaayman yes they can, but why isn't it the default behavior? Or should my snippet or @meteorlxy's one be added to the documentation? Because I think it's a behavior a lot of people would like to have.

**Comment by jasonsaayman** [comment]
@KoltesDigital yeah may be worth exploring, as it would make sense to allow it or make it more configurable. i just think we may have not done it it in the past due to the amount of internals that are in there. i will reopen and add to triage again, thanks for replying after like 2 years
